File: zad_9.py
from math import exp, cos

def simpson_rule(func, a, b, nseg):
    if nseg % 2 == 1:
        nseg += 1
    dx = (b - a) / nseg
    total = func(a) + func(b)
    for i in range(1, nseg, 2):
        total += 4 * func(a + i * dx)
    for i in range(2, nseg, 2):
        total += 2 * func(a + i * dx)
    return total * dx / 3

a_const = 1
b_const = 2

def func(x):
    return exp(-a_const * x) * cos(b_const * x)

print("Using Simpson's method")
print("Integrating function: f(x) = exp(-a*x) * cos(b*x), a = 1, b = 2")

a = 0
b = 50  
nseg = 2
eps = 1e-7

int_1 = simpson_rule(func, a, b, nseg)
int_2 = simpson_rule(func, a, b, nseg * 2)

while abs(int_1 - int_2) > eps:
    nseg *= 2
    int_1 = simpson_rule(func, a, b, nseg)
    int_2 = simpson_rule(func, a, b, nseg * 2)

print("\nAnswer: I =", int_2)
print("Number of segments:", nseg * 2)

File: zad_9.cpp
#include <iostream>
#include <cmath>
using namespace std;

double a_const = 1.0;
double b_const = 2.0;

double func(double x) {
    return exp(-a_const * x) * cos(b_const * x);
}

double simpson(double (*f)(double), double a, double b, int nseg) {
    if (nseg % 2 == 1) nseg++;
    double h = (b - a) / nseg;
    double result = f(a) + f(b);
    for (int i = 1; i < nseg; i += 2)
        result += 4 * f(a + i * h);
    for (int i = 2; i < nseg; i += 2)
        result += 2 * f(a + i * h);
    return result * h / 3.0;
}

int main() {
    double a = 0, b = 50;
    int nseg = 2;
    double eps = 1e-7;

    double int1 = simpson(func, a, b, nseg);
    double int2 = simpson(func, a, b, nseg * 2);

    while (fabs(int1 - int2) > eps) {
        nseg *= 2;
        int1 = simpson(func, a, b, nseg);
        int2 = simpson(func, a, b, nseg * 2);
    }

    cout << "Integral 9 Result: " << int2 << endl;
    cout << "Segments used: " << nseg * 2 << endl;
    return 0;
}